In [41]:
import numpy as np
import pandas as pd

In [42]:
def printM3(matrix3, text):
    print(f"\n{text}:\n[")
    for i in range(len(matrix3)):
        print(matrix3[i])
    print("]")

In [43]:
def calcFoward(values, weights):
    #print(f"values:{values}\nweights:{weights}")
    result = 0.0
    for i in range(len(weights)-1):
        result += values[i] * weights[i]
    result += 1 * weights[-1]
    #print(f"\ncalcFoward:{result}")
    return result

def calcCost(value, ideal):
    result = (ideal - value)**2/2
    #print(f"Cost:{result}")
    return result

def calcdETotalGo(value, ideal):
    result = -(ideal - value)
    #print(f"dETotal:{result}")
    return result

In [44]:
def relu(value):
    if(value < 0):
        value = 0
    print(f"\relu:{value}")
    return value

def drelu(value):
    if(value > 0):
        return value
    else:
        return 0

def linear(value):
    return value

def dlinear(value):
    return 1

def sigmoid(value):
    try:
        value = (1/(1 + np.exp(-value)))
    except RuntimeWarning:
        value = np.finfo(np.float64).tiny
    return value

def dsigmoid(value):
    value = value * (1 - value)
    return value

In [45]:
class Layer:
    def __init__(self,layer_size, neuron_size, activation_function, dactivation_function):
        self.neurons = np.random.rand(layer_size, neuron_size+1)
        #matrix = np.array(weights)
        #self.neurons = matrix
        print(f"\nNEURONS:{self.neurons}")
        self.activation_function = activation_function
        self.dactivation_function = dactivation_function
    
    def updateWeights(self, update_values, alpha):
        #print(f"Udate Weights\nUpdate Values:{update_values}\nAlpha:{alpha}")
        for i in range(len(self.neurons)):
            #print(f"\nNeuron {i}")
            for j in range(len(self.neurons[i])):
                #print(f"\nWeight {j}")
                #print(f"\nValue:{update_values[i][j]}")
                #print(f"Equation\n{self.neurons[i][j]} - {alpha} * {update_values[i][j]}")
                self.neurons[i][j] = self.neurons[i][j] - alpha * update_values[i][j]


    def foward(self,values):
        layer_a = []
        layer_z = []
        #print(f"\nValues:{values}")
        for i in range(len(self.neurons)):
            #print(f"\nNeuron:{i}")
            z = calcFoward(values,self.neurons[i])
            a = self.activation_function(z)
            #print(f"a:{a}\nz:{z}")
            layer_a.append(a)
            layer_z.append(z)
            
        #print(f"\nLayer A:{layer_a}\nLayer Z:{layer_z}")
        return np.array(layer_a),np.array(layer_z)
    
    def backpropagation(self, a0, a1, z0, ideals):
        #print(f"A0:{a0}\nA1:{a1}\nZ:{z0}\nIdeals:{ideals}")
        result = []
        aux = []
        for i in range(len(self.neurons)):
            neuron = []
            #print(f"\nNeuron:{i}")
            dEg = calcdETotalGo(a0[i],ideals[i])
            #print(f"dEg:{dEg}")
            dgouo = self.dactivation_function(a0[i])
            #print(f"dgouo:{dgouo}")
            #pesos
            for j in range(len(self.neurons[i])-1):
                #print(f"\nWeight:{j}")
                dgh = a1[j]
                #print(f"dgh:{dgh}")
                dEw = dEg * dgouo * dgh
                #print(f"dEw:{dEw}")
                neuron.append(dEw)
            #bias
            #print("\nBias:")
            dEw = dEg * dgouo
            #print(f"dEw:{dEw}")
            neuron.append(dEw)
            #print(f"Fix Result:{neuron}")
            result.append(np.array(neuron))
            aux.append(dEw)
            #print(f"aux {i}:\n{aux}")
        return np.array(result), np.array(aux)

    def backpropagationHidden(self, dEu, a0, a1, previous_layer):
        #print(f"\nA0:{a0}\nA1:{a1}\ndEu:{dEu}\nWeights:{self.neurons}\nPrevious Layer:{previous_layer}")
        result = []
        next = []
        for i in range(len(self.neurons)):
            neuron = []
            #print(f"\nNeuron:{i}")
            dghuh = self.dactivation_function(a0[i])
            #print(f"dghuh:{dghuh}")
            true_dEg = []
            for j in range(len(previous_layer)):
                for k in range(len(dEu)):
                    #print(f"\nCalc dEg:{j}")
                    dEg = previous_layer[j][i] * dEu[k] 
                    #print(f"W:{previous_layer[j][i]}\ndEu:{dEu[j]}\ndEg:{dEg}")
                    true_dEg.append(dEg)
            true_dEg = sum(true_dEg)
            next.append(true_dEg)
            #print(f"\nTrue dEg:{true_dEg}")
            #pesos
            for k in range(len(self.neurons[i])-1):
                #print(f"\nWeight:{k}")
                dgh = a1[k]
                #print(f"dgh:{dgh}")
                dEw = true_dEg * dghuh * dgh
                #print(f"\ndEw:{dEw}")
                neuron.append(dEw)
            #bias
            #print(f"\nBias:")
            dEw = true_dEg * dghuh
            #print(f"\ndEw:{dEw}")
            neuron.append(dEw)
            result.append(np.array(neuron))


        return np.array(result),np.array(next)


    def relat(self):
        print(self.neurons)

In [46]:
class Network:
    def __init__(self):
        self.layers = []

    def appendLayer(self, layer_size, neuron_size, activation_function, dactivation_function):
        self.layers.append(Layer(layer_size, neuron_size, activation_function, dactivation_function))

    def foward(self,values):
        #print(f"\nFirst_Values:{values}")
        layer_a = [values]
        layer_z = []
        for i in range(len(self.layers)):
            #print(f"\nFoward:{i}")
            values,z = self.layers[i].foward(values)
            layer_a.append(values)
            layer_z.append(z)
        return layer_a,layer_z
                
    def backpropagation(self, all_a, all_z, ideals):
        #print(f"\nBack First")
        #printM3(all_a,"All A")
        #printM3(all_z,"All Z")
        #print(f"\nIdeals:{ideals}")
        result = []
        next = []
        for i in range(len(self.layers)):
            if i == 0:
                #print(f"\nBack")
                aux, next = self.layers[-(i+1)].backpropagation(all_a[-(i+1)], all_a[-(i+2)], all_z[-(i+1)], ideals)
                result.append(aux)
                #print(next)
            else:
                #print(f"\nBack Hidden {i}")
                aux, next = self.layers[-(i+1)].backpropagationHidden(next, all_a[-(i+1)], all_a[-(i+2)], self.layers[-i].neurons)
                result.append(aux)
        return result
        
    def calcError(self, output, ideals):
        result = 0
        #print(f"Ideals:{ideals}\nOut:{output}")
        for i in range(len(output)):
            result = result + calcCost(output[i], ideals[i])
        #print(f"Total cost:{result}")
        return result
    
    def updateWeights(self, update_values, alpha):
        for i in range(len(self.layers)):
            self.layers[-(i+1)].updateWeights(update_values[i],alpha)

    def train(self, input, ideals, alpha, epoch):
        print("Training")
        acumulated = 0
        for i in range(epoch):
            error = []
            for j in range(len(input)):
                all_a, all_z = self.foward(input[j])
                #print(f"\nAll Activations:\n{all_a}")
                if(i <= 2 or i >= 998):
                    error.append(self.calcError(all_a[-1],ideals[j]))
                fix = self.backpropagation(all_a, all_z, ideals[j])
                #print(f"fix:{fix}")
                if type(acumulated) is int:
                    acumulated = fix
                else:
                    for k in range(len(acumulated)):
                        for l in range(len(acumulated[k])):
                            acumulated[k][l] = acumulated[k][l] + fix[k][l]
                #print(f"\nacc {j}:\n{acumulated}")
            if(i <= 2 or i >= epoch-2):
                error = sum(error)
                error = error / len(input)
                print(f"Total Error Epoch {i}: {error}")
            #print(f"Acumulated sum:{acumulated}")
            for k in range(len(acumulated)):
                for l in range(len(acumulated[k])):
                    acumulated[k][l] = acumulated[k][l] / len(input)
            #print(f"Acumulated div:{acumulated}")
            self.updateWeights(acumulated,alpha)

    def test(self, input, ideals):
        print("Testing")
        error = []
        for j in range(len(input)):
            all_a, all_z = self.foward(input[j])
            error.append(self.calcError(all_a[-1],ideals[j]))
        error = sum(error)
        error = error / len(input)
        print(f"Total Error:{error}")
    
    def relat(self):
        for i in range(len(self.layers)):
            print(f"\nLayer{i}:")
            self.layers[i].relat()


In [47]:
network = Network()
network.appendLayer(8,4,sigmoid,dsigmoid)
network.appendLayer(4,8,sigmoid,dsigmoid)
network.appendLayer(1,4,sigmoid,dsigmoid)


NEURONS:[[0.29751782 0.19450498 0.55320811 0.989759   0.94000046]
 [0.04786209 0.25989237 0.06132024 0.40202024 0.91497649]
 [0.20806137 0.42566086 0.28683948 0.60648868 0.15091016]
 [0.71925036 0.6624696  0.85477155 0.42197915 0.22104436]
 [0.52496676 0.17207643 0.94335375 0.24885774 0.38891095]
 [0.46721426 0.92492928 0.14848644 0.91786748 0.43208742]
 [0.32407054 0.09935384 0.30017519 0.06421489 0.66969776]
 [0.57804906 0.580806   0.05798437 0.42453903 0.76306453]]

NEURONS:[[0.46382833 0.54238368 0.22416321 0.72142642 0.91993547 0.13755571
  0.81503319 0.30157633 0.44029021]
 [0.99583545 0.47936873 0.77774964 0.52099784 0.19832643 0.22182506
  0.86878377 0.09924862 0.85347978]
 [0.30203188 0.32601103 0.26522486 0.87808554 0.59811301 0.8040257
  0.89452186 0.03318196 0.94807431]
 [0.56315852 0.91696105 0.952711   0.22153005 0.08965613 0.18677093
  0.1742667  0.08239901 0.92405446]]

NEURONS:[[0.56166194 0.16293548 0.46136873 0.07486988 0.79460293]]


In [48]:
def prepare_Y(y):
    mapping = {
        "Iris-setosa": 0,
        "Iris-versicolor": 1,
        "Iris-virginica": 2,
    }

    return y.map(mapping)

# Definição do dataframe iris Flowers
df = pd.read_csv('IRIS.csv')


X = df.drop(columns=['species']).to_numpy()

Y = prepare_Y(df['species']).to_numpy()

clean_input = []
for i in range(len(X)):
    clean_input.append(np.array(X[i]))

Y = Y.tolist()

clean_all_ideals = []
for i in range(len(Y)):
    aux = [Y[i]]
    clean_all_ideals.append(aux)

#print(clean_all_ideals)

network.relat()

network.train(clean_input, clean_all_ideals, 0.5, 10000)

network.relat()


Layer0:
[[0.29751782 0.19450498 0.55320811 0.989759   0.94000046]
 [0.04786209 0.25989237 0.06132024 0.40202024 0.91497649]
 [0.20806137 0.42566086 0.28683948 0.60648868 0.15091016]
 [0.71925036 0.6624696  0.85477155 0.42197915 0.22104436]
 [0.52496676 0.17207643 0.94335375 0.24885774 0.38891095]
 [0.46721426 0.92492928 0.14848644 0.91786748 0.43208742]
 [0.32407054 0.09935384 0.30017519 0.06421489 0.66969776]
 [0.57804906 0.580806   0.05798437 0.42453903 0.76306453]]

Layer1:
[[0.46382833 0.54238368 0.22416321 0.72142642 0.91993547 0.13755571
  0.81503319 0.30157633 0.44029021]
 [0.99583545 0.47936873 0.77774964 0.52099784 0.19832643 0.22182506
  0.86878377 0.09924862 0.85347978]
 [0.30203188 0.32601103 0.26522486 0.87808554 0.59811301 0.8040257
  0.89452186 0.03318196 0.94807431]
 [0.56315852 0.91696105 0.952711   0.22153005 0.08965613 0.18677093
  0.1742667  0.08239901 0.92405446]]

Layer2:
[[0.56166194 0.16293548 0.46136873 0.07486988 0.79460293]]
Training
Total Error Epoch 0: 0.3

#esse era o principal teste, onde Network2 permitia criar a rede com o peso dos neuronios já definidos
network2 = Network2()
network2.appendLayer(2,2,sigmoid,dsigmoid,[[0.15, 0.2, 0.35], [0.25, 0.3, 0.35]])
network2.appendLayer(2,2,sigmoid,dsigmoid,[[0.4, 0.45, 0.6], [0.5, 0.55, 0.6]])

input = [[1,1],[0.05,0.1],[1,0],[0,0]]
all_ideals = [[0,1],[0.01,0.99],[1,0],[0,1]]

clean_input = []
for i in range(len(input)):
    clean_input.append(np.array(input[i]))

clean_all_ideals = np.array(all_ideals)

network2.train([clean_input[1]], [clean_all_ideals[1]], 0.5, 10000)